### PR TITLE
Yamcs fault management

### PR DESCRIPTION
**Pull request by nikhilmandlik**
Closes https://github.com/akhenry/openmct-yamcs/issues/140

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [ ] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue?

### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Changes appear not to be breaking changes?
* [ ] Appropriate unit tests included?
* [x] Code style and in-line documentation are appropriate?
* [x] Commit messages meet standards?
* [x] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [ ] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

File: src/providers/fault-mgmt-providers/fault-action-provider.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { FAULT_MANAGEMENT_ALARMS, FAULT_MANAGEMENT_TYPE } from './fault-mgmt-constants';
+
+export default class HistoricalFaultProvider {
+    constructor(configuration) {
+        this.configuration = configuration;
+    }
+
+    shelveFault(fault) {
+        console.log('shelveFault', fault, this.configuration);
```
Comment (nikhilmandlik) [inline src/providers/fault-mgmt-providers/fault-action-provider.js:9]: Currently WIP, need to `POST` on yamcs url from config

File: package.json [hunk 2]
```diff
@@ -4,26 +4,29 @@
...
     "start": "NODE_ENV=dev webpack-dev-server",
     "prepare": "npm run build:dist"
   },
   "keywords": [
-    "openmct",
     "yamcs"
   ],
   "author": "National Aeronautics and Space Administration",
   "license": "Apache-2.0",
+  "dependencies": {
```
Comment (akhenry) [inline package.json:20]: This will cause Open MCT to be fetched and compiled whenever openmct-yamcs is included, which we don't want. Open MCT is a peer dependency of openmct-yamcs, not a direct dependency.

File: src/plugin.js [hunk 3]
```diff
@@ -42,18 +46,35 @@ export default function installYamcsPlugin(configuration) {
...
-            configuration.yamcsWebsocketEndpoint,
+        const realtimeTelemetryProvider = new RealtimeProvider(
+            yamcsWebSocket,
             configuration.yamcsInstance
         );
-        openmct.telemetry.addProvider(realtimeProvider);
-        realtimeProvider.connect();
+        openmct.telemetry.addProvider(realtimeTelemetryProvider);
+
+        openmct.faults.addHistoricalProvider(new HistoricalFaultProvider(
```
Comment (akhenry) [inline src/plugin.js:64]: Let's use named arguments. eg.
``` JavaScript
openmct.faults.addProvider(new YamcsFaultProvider({
    historicalEndpoint: configuration.yamcsHistoricalEndpoint,
    yamcsInstance: configuration.yamcsInstance
});
```

File: src/providers/historical-telemetry-provider.js [hunk 4]
```diff
@@ -145,7 +145,7 @@ export default class YamcsHistoricalTelemetryProvider {
     isImagery(domainObject) {
         let metadata = this.openmct.telemetry.getMetadata(domainObject);
 
-        return metadata.valuesForHints(['image']).length !== 0;
+        return metadata && metadata.valuesForHints(['image']).length !== 0;
```
Comment (akhenry) [inline src/providers/historical-telemetry-provider.js:148]: Was there an issue with metadata not being defined? It should always be defined.

File: src/providers/yamcs-web-socket.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
+const FALLBACK_AND_WAIT_MS = [1000, 5000, 5000, 10000, 10000, 30000];
+
+export default class YamcsWebSocket {
```
Comment (akhenry) [inline src/providers/yamcs-web-socket.js:3]: 👍 Good stuff!

File: src/providers/realtime-provider.js [hunk 6]
```diff
@@ -99,166 +88,119 @@ export default class RealtimeProvider {
...
-        }, FALLBACK_AND_WAIT_MS[this.currentWaitIndex]);
-
-        if (this.currentWaitIndex < FALLBACK_AND_WAIT_MS.length - 1) {
-            this.currentWaitIndex++;
-        }
-    }
-
-    sendOrQueueMessage(request) {
-        if (this.connected) {
+    _flushQueue() {
```
Comment (akhenry) [inline src/providers/realtime-provider.js:96]: The queuing logic needs to be moved into the yamcs-web-socket, because the realtime fault provider is going to need this too. This is needed when subscription requests arrive when the websocket is not connected (either because it's still handshaking, or it has dropped out - which happens regularly).

File: src/providers/yamcs-web-socket.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+        return () => {
+            this._oncloseSubscribers.delete(oncloseCallback);
+        };
+    }
+
+    onerror(callback) {
+        const onerrorCallback = { callback };
+
+        this.onerrorSubscribers.add(onerrorCallback);
```
Comment (akhenry) [inline src/providers/yamcs-web-socket.js:66]: Typo? should be `this._onerrorSubscribers`.

File: src/providers/yamcs-web-socket.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        const onopenCallback = { callback };
+
+        this._onopenSubscribers.add(onopenCallback);
+
+        return () => {
+            this._onopenSubscribers.delete(onopenCallback);
+        };
+    }
+
+    sendMessage(message) {
```
Comment (akhenry) [inline src/providers/yamcs-web-socket.js:93]: This needs to queue messages if the websocket is not connected right now. The realtime-provider has this logic already, it should be moved here.

**Review by akhenry — CHANGES_REQUESTED** [review]
Looking really good. I think the refactoring of websocket handling especially! We need to be really careful with websockets to make sure we are maintaining the error handling that it currently does, there was a lot of hard work involved in making websockets stable and reliable.

File: src/plugin.js [hunk 9]
```diff
@@ -42,18 +46,35 @@ export default function installYamcsPlugin(configuration) {
...
+            configuration.yamcsHistoricalEndpoint,
+            configuration.yamcsInstance
+        ));
+
+        openmct.faults.addRealtimeProvider(new RealtimeFaultProvider(
+            yamcsWebSocket,
+            configuration.yamcsInstance
+        ));
+
+        openmct.faults.addFaultActionProvider(new FaultActionrovider(
```
Comment (shefalijoshi) [inline src/plugin.js:74]: Typo in `FaultActionrovider` - missing `P`

File: package.json [hunk 10]
```diff
@@ -4,26 +4,29 @@
...
     "test": "echo \"Error: no test specified\" && exit 1",
     "lint": "eslint src example",
     "lint:fix": "eslint src example --fix",
     "build:dist": "NODE_ENV=production webpack",
-    "prepublish": "npm install openmct@unstable --no-save",
     "start": "NODE_ENV=dev webpack-dev-server",
     "prepare": "npm run build:dist"
   },
   "keywords": [
-    "openmct",
```
[the remark this answers is not in the document]
Comment (nikhilmandlik) [reply package.json]: by mistake

File: src/providers/historical-telemetry-provider.js (hunk 4 above)
Comment (nikhilmandlik) [reply src/providers/historical-telemetry-provider.js]: I think it was when I was trying to use old providers for fault and fault mgmt does not have metadata

File: webpack.config.js [hunk 11]
```diff
@@ -33,23 +33,31 @@ const WEBPACK_CONFIG = {
         } else {
```
Comment (akhenry) [inline webpack.config.js:33]: @jvigliotta https://github.com/akhenry/openmct-yamcs/pull/131 has the same changes, this file can be reverted.

File: package-lock.json [hunk 12]
```diff
@@ -1,6333 +0,0 @@
-{
```
Comment (akhenry) [inline package-lock.json:1]: @jvigliotta Don't delete please. Revert this change.

**Review by akhenry — CHANGES_REQUESTED** [review]

File: src/providers/yamcs-web-socket.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+            this._onmessageSubscribers.forEach(onmessageSubscriber => onmessageSubscriber.callback(event));
+        };
+
+        this._socket.onerror = (error) => {
+            console.error(error);
+            console.warn("Websocket error, attempting reconnect...");
+            this._connected = false;
+            this._reconnect();
+
+            this.onerrorSubscribers.forEach(onerrorSubscriber => onerrorSubscriber.callback(error));
```
Comment (akhenry) [inline src/providers/yamcs-web-socket.js:43]: Typo. Should be `this._onerrorSubscribers`.

File: src/providers/yamcs-web-socket.js [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+        this._onopenSubscribers.add(onopenCallback);
+
+        return () => {
+            this._onopenSubscribers.delete(onopenCallback);
+        };
+    }
+
+    sendMessage(message) {
+        if (!this._connected) {
```
Comment (akhenry) [inline src/providers/yamcs-web-socket.js:96]: No. Get rid of this method and replace it with [the original `sendOrQueueMessage`](https://github.com/akhenry/openmct-yamcs/blob/7916c86ea7bac5037f8c4c5253857aad58bf3739/src/providers/realtime-provider.js#L130). Right now you're queuing in a method called `sendMessage`, and sending in a method called `_sendOrQueue`.

File: src/providers/yamcs-web-socket.js [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+    _sendMessage(message) {
+        this._socket.send(message);
+    }
+
+    _sendOrQueueMessage(message) {
+        try {
+            this._sendMessage(message);
+        } catch (error) {
+            console.error(error);
```
Comment (akhenry) [inline src/providers/yamcs-web-socket.js:142]: On error this should reconnect, [as in the original code](https://github.com/akhenry/openmct-yamcs/blob/7916c86ea7bac5037f8c4c5253857aad58bf3739/src/providers/realtime-provider.js#L130). As I said, __we have to get this right__. The stability of our WebSocket connections has been hard won over the years, with a lot of trial and error.

**Review by akhenry — CHANGES_REQUESTED** [review]

File: example/index.js [hunk 16]
```diff
@@ -86,5 +86,7 @@ const openmct = window.openmct;
         openmct.install(openmct.plugins.Notebook());
         openmct.install(openmct.plugins.LADTable());
         openmct.install(openmct.plugins.ClearData(['table', 'telemetry.plot.overlay', 'telemetry.plot.stacked']));
+
+        openmct.install(openmct.plugins.FaultManagement());
```
Comment (jvigliotta) [inline example/index.js:90]: Had to rename this, it was "FaultManagementPlugin" which was incorrect.

**Review by akhenry — APPROVED** [review]
